collections/max_by.ts

```diff
@@ -0,0 +1,40 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * Returns the first element that is the largest value of the given function or undefined if there are no elements.
+ *
+ * Example:
+ *
+ * ```ts
+ * import { maxBy } from "./max_by.ts";
+ * import { assertEquals } from "../testing/asserts.ts";
+ *
+ * const people = [
+ *     { name: 'Anna', age: 34 },
+ *     { name: 'Kim', age: 42 },
+ *     { name: 'John', age: 23 },
+ * ];
+ *
+ * const personWithMaxAge = maxBy(people, i => i.age);
+ *
+ * assertEquals(personWithMaxAge, { name: 'Kim', age: 42 });
+ * ```
+ */
+export function maxBy<T>(
+  collection: Array<T>,
+  selector: ((el: T) => number | string),
```

Comment:
I think this typing is too narrow - it allows a function that **sometimes** a `number` and **sometimes** returns a `string`.

You can do something like `((el: T) => number) | ((el: T) => string)` instead. I would add an overload to the actual exposed signature though, that is more readable and nicer to use.